Entity endpoint Controller and Integration test

rma-kayla opened the pull request:
Implemented new Entity controller that exposes all CRUD methods. 
Created accompanying integration tests to ensure the endpoints work correctly and return the appropriate error responses.

rma-kayla committed:
CDA-54 Create EntityController:
Initial implementation of controller and CRUD basics.

rma-rripken committed:
Bugfix/cda 45 ts vertical datum (#1344)

fixes https://github.com/USACE/cwms-data-api/issues/1102

---------

Co-authored-by: Bryson Spilman <bryson@rmanet.com>

MikeNeilson committed:
Update treafik for latest docker. (#1493)

Previous version of traefik couldn't communicate to latest version of
docker. Update traefik so the compose file works again.

DanielTOsborne committed:
Enhancements/blob clob query (#1483)

Add blob-id or clob-id query parameter, respectively, to all required
methods, so ones with invalid path characters can be referenced.

This is required for
https://github.com/HydrologicEngineeringCenter/cwms-cli/issues/72

rma-kayla committed:
CDA-54 Implemented EntityController and Integration test:
- Refactored and finalized CRUD methods.
- Verified request validation and error handling
- Added route integration and comprehensive integration tests.

rma-kayla commented:
@rma-bryson this PR is ready for your review

rma-kayla committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/CDA-54-Entity-Endpoint

rma-bryson commented on cwms-data-api/src/main/java/cwms/cda/api/EntityController.java:
```diff
@@ -0,0 +1,253 @@
+                return;
+            }
+            EntityDao dao = new EntityDao(dsl);
+            // verify entity exists before updating
```
I should have done this check in the dao. Can you move the check into the `updateEntity` method? You shouldn't need to catch any exception, it will throw a NotFoundException automatically and should then return the appropriate status - should verify this with integration test (see other comment).

rma-kayla replied:
Moved the check if entity exists before updating into the dao

rma-bryson commented on cwms-data-api/src/test/java/cwms/cda/api/EntityControllerTestIT.java:
```diff
@@ -0,0 +1,294 @@
+        .assertThat()
+            .statusCode(is(HttpServletResponse.SC_OK));
+
+        // GET to confirm updated field persisted
```
Add an update-with-non-existing-entity-id check here with a bogus entity id to make sure its returning the not-found status.

rma-kayla replied:
I have a test for this outside of this test.  I'll rename it to make it more explicit.
// Entity ID in the URL must match the id.name in the request body
    @Test
    void update_mismatched_entity_id_bad_request()

rma-bryson commented on cwms-data-api/src/main/java/cwms/cda/api/EntityController.java:
```diff
@@ -0,0 +1,253 @@
+import static cwms.cda.data.dao.JooqDao.getDslContext;
+
+public class EntityController implements CrudHandler {
+    public static final String TAG = "Entity";
```
the TAG and LOGGER constants can be private

rma-kayla replied:
made Tag private, and removed unused logger

rma-bryson commented on cwms-data-api/src/test/java/cwms/cda/api/EntityControllerTestIT.java:
```diff
@@ -0,0 +1,294 @@
+
+    // Show simple filtering works with getAll and parent entity id only
+    @Test
+    void getAll_with_parent_filter_returns_200_empty_list_ok() {
```
Add a test for the match null parents flag. 
1. Create a entity with a null parent
2. GET request with flag set to true -> should get it back
3. GET request with flag set to false -> shouldn't get it back. 
4. delete entity

rma-kayla replied:
created test to verify match-null-parents functionality

rma-kayla committed:
CDA-54 refactored EntityController
- moved update verification into DAO
- cleaned up variables
- added match null parents test to integration test
